### PR TITLE
Code to parse all latest scrapes and save locally

### PR DESCRIPTION
**Pull request by lizgzil**
# Description

This PR adds code to run the parser with the latest scraped files in one go.
I add a description of how to do this in the README, and add to new files:
`utils/export_env` which exports the latest file names as environment variables, and
`parse_latest.py` which parses the latest scraped files, and outputs the parsed references and matches locally.

## Type of change

- [x] :sparkles: New feature

# How Has This Been Tested?

I have manually checked that the latest file names in S3 are the same as my environmental variables.
I have started to run the code and it does start to save the outputs as I expected it to, however I didn't run it until it was complete.

# Checklist:

- [x] My code follows the style guidelines of this project (pep8 AND pyflakes)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] If needed, I changed related parts of the documentation
- [ ] I included tests in my PR
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] If my PR aims to fix an issue, I referenced it using `#(issue)`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -60,6 +60,18 @@ python3 $DIR/main.py \
 
 If the `scraper_file`, `references_file`, `model_file`, and `vectorizer_file` arguments are to S3 locations then make sure these start with `s3://`, otherwise file names are assumed to be locally stored. If the `output_url` argument is to a local location, then make sure it begins with `file://`, otherwise it is assumed to be from a database.
 
+### Method 4.
+
+If you would like to run the parser for the latest scraped files and to save the output locally, then you will first need to get the latest file names by running
+```
+chmod u+x utils/export_env
+source utils/export_env
```
Comment (nsorros) [inline README.md]: you do `chmod u+x utils/export_env` so that you do need to do `source` so pick one.

Comment (lizgzil) [reply README.md]: I'm a bit confused because when I did `chmod u+x utils/export_env` it just changed `export_env` to an executable file, but didn't save the variables, and then when I ran `source utils/export_env` I had the files in my environment variables. Does it work for you if you just do `source utils/export_env`?

Comment (nsorros) [reply README.md]: and plus i am not sure you need to do every time

File: README.md [hunk 2]
```diff
@@ -60,6 +60,18 @@ python3 $DIR/main.py \
...
+```
+And then run
+`python parse_latest.py`
```
Comment (nsorros) [inline README.md]: initially what i was thinkinh is that you would make the environment variables available so you can run python main.py with the arguments being available from export_env but i see that you are trying to run the parser for all organisations. hmm...

File: parse_latest.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+gov_uk = os.environ['GOVUK_S3_URL']
+msf = os.environ['MSF_S3_URL']
+nice = os.environ['NICE_S3_URL']
+parliament = os.environ['PARLIAMENT_S3_URL']
+unicef = os.environ['UNICEF_S3_URL']
+who_iris = os.environ['WHO_S3_URL']
+
+latest_names = {
+				'gov_uk': gov_uk,
```
Comment (nsorros) [inline parse_latest.py]: there is something wrong here with the tabs or spaces

File: parse_latest.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import os
+from settings import settings
+
+gov_uk = os.environ['GOVUK_S3_URL']
+msf = os.environ['MSF_S3_URL']
+nice = os.environ['NICE_S3_URL']
+parliament = os.environ['PARLIAMENT_S3_URL']
+unicef = os.environ['UNICEF_S3_URL']
+who_iris = os.environ['WHO_S3_URL']
```
Comment (nsorros) [inline parse_latest.py]: it is a good idea to check whether these variables are not empty

File: parse_latest.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import os
```
Comment (nsorros) [inline parse_latest.py]: It will also be good to write what this file is for in the beggining

File: parse_latest.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	# The directory will just have the name of the organisation and the date of scrape
+	folder_name = organisation + latest_name_organisation[-13:-5]
+	dir_name = './tmp/parser-output/{}'.format(folder_name)
+
+	if not os.path.exists(dir_name):
+		os.mkdir(dir_name)
+
+	run_script = 'python3 main.py \
+	 	--scraper-file "s3://datalabs-data/{}" \
+	 	--references-file "s3://datalabs-data/wellcome_publications/uber_api_publications.csv" \
```
Comment (nsorros) [inline parse_latest.py]: Also i remind you that this task is about running against a given set of publications like the one from UKRI or from the population health team. So you need an argument to give the publications and you can use the wellcome ones that the parser uses as the default one.

File: parse_latest.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+who_iris = os.environ['WHO_S3_URL']
+
+latest_names = {
+				'gov_uk': gov_uk, 
+				'msf': msf,
+				'nice': nice, 'parliament':parliament, 'unicef':unicef,'who_iris':who_iris
+				}
+
+# Use these to run the parser for the organisations with regex:
+organisation_regex = list(settings._regex_dict.keys())
```
[the remark this answers is not in the document]
Comment (lizgzil) [reply parse_latest.py]: I wanted to make it only process the organisations which we have regexes for. How it is at the moment means we don't have to manually check which ones have regex.

File: parse_latest.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+for organisation in organisation_regex:
+
+	latest_name_organisation = latest_names[organisation]
+
+	# The directory will just have the name of the organisation and the date of scrape
+	folder_name = organisation + latest_name_organisation[-13:-5]
```
[the remark this answers is not in the document]
Comment (lizgzil) [reply parse_latest.py]: I will make the comment clearer (the date of scrape is the -13:-5th characters of latest_name_organisation)

File: utils/export_env [hunk 9]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+export GOVUK_S3_URL=$(aws s3 ls datalabs-data/scraper-results/gov_uk/ --recursive | sort | tail -n 1 | awk '{print $4}')
```
[the remark this answers is not in the document]
Comment (SamDepardieu) [reply utils/export_env]: From the scraper code:
```
def _store_in_thread(self, data_file):
        """This method will try to upload the file to S3.
        """
        data_file.seek(0)
        date_tag = datetime.now().strftime('%Y%m%d')
        try:
            last_content = self._get_last_result()
            content = last_content + data_file.read().decode()
            filename = "{folders}/{file_datetag}.json".format(
                folders=self.s3_file,
                file_datetag=date_tag
            )
            self.s3.put_object(
                Body=content,
                Bucket=self.s3_bucket,
                Key=filename
            )
        except ClientError as e:
self.logger.error('Couldn\'t upload the json file to s3: %s', e)
```
In theory, the last file should contain every publication seen by the parser

File: parse_latest.py (hunk 3 above)
Comment (lizgzil) [reply parse_latest.py]: will fix

File: parse_latest.py (hunk 5 above)
Comment (lizgzil) [reply parse_latest.py]: I can write `# So we can get environmental variables` next to this, but we don't usually comment on importing `os`, is this what you mean here or have I misunderstood?

File: parse_latest.py (hunk 4 above)
Comment (lizgzil) [reply parse_latest.py]: 👍

File: parse_latest.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nsorros) [reply parse_latest.py]: it would be more preferable to extract the date differently because it seems that if a character changes at the moment this code stops working, what is an example file name?

File: parse_latest.py (hunk 5 above)
Comment (nsorros) [reply parse_latest.py]: misunderstood. i mean write what this script does in the first line or somewhere.

File: parse_latest.py (hunk 6 above)
Comment (lizgzil) [reply parse_latest.py]: 👍

File: parse_latest.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (lizgzil) [reply parse_latest.py]: An example of `latest_name_organisation` is `scraper-results/who_iris/20181204.json`, all the files are 8 character dates. I've changed this to `os.path.basename(latest_name_organisation)` which is nicer I think.

File: parse_latest.py (hunk 5 above)
Comment (lizgzil) [reply parse_latest.py]: haha 🤦‍♀️

File: parse_latest.py (hunk 6 above)
Comment (lizgzil) [reply parse_latest.py]: with regard to this I've changed things a bit to allow this to be an argument

**Comment by lizgzil** [comment]
@nsorros I've addressed your comments in the latest commit :)
a change in indentation makes it look bigger than it is!

File: parse_latest.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+	Argument is the references file you want to match the parsed references against
+	If local this should start with 'file://'
+	"""
+	try:
+		references_file = sys.argv[1]
+	except:
+		references_file = "s3://datalabs-data/wellcome_publications/uber_api_publications.csv"
+
+	# Use these to run the parser for the organisations with regex:
+	organisation_regex = list(settings._regex_dict.keys())
```
Comment (nsorros) [inline parse_latest.py]: maybe `organisations_with_regex` since it is more than one and this is not a regex.

File: parse_latest.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+gov_uk = os.environ.get('GOVUK_S3_URL', None)
+msf = os.environ.get('MSF_S3_URL', None)
+nice = os.environ.get('NICE_S3_URL', None)
+parliament = os.environ.get('PARLIAMENT_S3_URL', None)
+unicef = os.environ.get('UNICEF_S3_URL', None)
+who_iris = os.environ.get('WHO_S3_URL', None)
+
+latest_names = {
+'gov_uk': gov_uk,
```
Comment (nsorros) [inline parse_latest.py]: add a tab or spaces here

File: README.md (hunk 1 above)
Comment (hblanks) [reply README.md]: Put short, the `chmod` should go.

Put longer: `chmod` makes it so you can do things like run `utils/export_env` (note we'd usually name this `export_env.sh`). In this case, running such a command would do little, because the shell passes this file to the kernel to execute, the kernel sees it's a shell script and starts off a new shell process, the script runs and sets environment variables for the new shell process, and then the new process exits.

And we're back to the user's interactive shell, with no new environment.

So, only `source` applies here.

Also note that git tracks whether files are executable or not so if you did need to make a file executable you'd do it once in your repo & commit the file.

**Comment by nsorros** [comment]
Let's park this until Monday or until @lizgzil is back and have a conversation with @hblanks on Monday to decide best way to proceed.

**Comment by lizgzil** [comment]
@hblanks I get a `NameError: name 'pool_map' is not defined` error when I run the latest code, but not sure why, any ideas?

**Comment by hblanks** [comment]
I'm not sure why offhand, but can take a look tomorrow if you like.

File: parse_latest.py [hunk 12]
```diff
@@ -8,59 +8,105 @@
...
 """
 
+from argparse import ArgumentParser
+from urllib.parse import urlparse
 import os
 import sys
+import boto3
+
 from settings import settings
+from refparse import parse_references
```
Comment (nsorros) [inline parse_latest.py]: I am under the impression you are not using this and instead you are still running `python main` with `os.system` which i think was not the intention

File: parse_latest.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    bucket_name, prefix = u.netloc, u.path[1:]
+
+    s3 = boto3.resource('s3')
+
+    my_bucket = s3.Bucket(bucket_name)
+
+    # Get the most recently scraped filename
+    key_name, obj = max(
+        (obj.key, obj) for obj in my_bucket.objects.filter(Prefix=prefix).all()
+    )
```
Comment (hblanks) [inline parse_latest.py]: this is kinda silly, but since have some edits to do below, so that you call `parse_references()` or such, maybe we can do a couple tiny edits here:

```
s3 = boto.resource('s3')  # no newline b/w the two, not that it really matters...
bucket = s3.Bucket(bucket_name)  # dropping the my_
```

File: README.md (hunk 1 above)
Comment (hblanks) [reply README.md]: We'll update the README, as we won't need `export_env` anymore.

File: utils/export_env (hunk 9 above)
[the remark this answers is not in the document]
Comment (hblanks) [reply utils/export_env]: Note we can drop this file; it's "OBE" as the PostgreSQL maintainers say. (Overcome by Events)

File: README.md (hunk 1 above)
Comment (lizgzil) [reply README.md]: 👍

File: parse_latest.py (hunk 12 above)
Comment (lizgzil) [reply parse_latest.py]: `main` is now called `refparse`, so we are using `parse_references`

File: parse_latest.py (hunk 13 above)
Comment (lizgzil) [reply parse_latest.py]: 👍

File: utils/export_env (hunk 9 above)
[the remark this answers is not in the document]
Comment (lizgzil) [reply utils/export_env]: 👍

**Comment by lizgzil** [comment]
@hblanks @nsorros this is working now without errors 🎉

File: README.md [hunk 14]
```diff
@@ -45,21 +45,40 @@ Once you're happy with your configuration, just run `python main.py`
...
 
-python3 $DIR/main.py \
+./refparse.py \
 	--scraper-file "s3://datalabs-data/scraper-results/msf/20190117.json" \
 	--references-file "match-references/MRC_Publications_Nov2018_JGHT_JHSRI.csv" \
 	--model-file "s3://datalabs-data/reference_parser_models/RefSorter_classifier.pkl" \
 	--vectorizer-file "s3://datalabs-data/reference_parser_models/RefSorter_vectorizer.pkl" \
-	--output-url "file:///tmp/parser-output/output_folder_name"
+	--output-url "file:///tmp/parser-output/output_folder_name" \
+    --num-workers None
```
Comment (nsorros) [inline README.md]: there is a tab problem here

File: refparse.py [hunk 15]
```diff
@@ -75,10 +75,10 @@ def get_file(file_str, file_type, get_scraped = False):
     return file
 
 def run_predict(scraper_file, references_file,
-                model_file, vectorizer_file,
-                pool_map, output_url):
+    model_file, vectorizer_file, pool_map, output_url):
```
Comment (nsorros) [inline refparse.py]: possibly a tab missing here?

File: refparse.py [hunk 16]
```diff
@@ -187,102 +282,33 @@ def run_predict(scraper_file, references_file,
...
+            parse_references_profile(args.scraper_file, args.references_file,
+                args.model_file, args.vectorizer_file, args.output_url)
         else:
-            run_predict(scraper_file, references_file,
-                        model_file, vectorizer_file, pool_map, output_url)
-
-        if pool is not None:
-            pool.terminate()
-            pool.join()
+            parse_references(args.scraper_file, args.references_file,
```
Comment (nsorros) [inline refparse.py]: the problem we have at the moment is this does not change the regex based on the organization information. so as discussed, create an issue about solving that long term and update the documentation. do that in a separate PR

**Comment by lizgzil** [comment]
@nsorros I've made the discussed changes :)

File: refparse.py (hunk 16 above)
Comment (lizgzil) [reply refparse.py]: i've added an issue, but added the temporary hack to the README in this PR. is that OK?

File: README.md [hunk 17]
```diff
@@ -45,31 +45,39 @@ Once you're happy with your configuration, just run `python main.py`
...
 If the `scraper_file`, `references_file`, `model_file`, and `vectorizer_file` arguments are to S3 locations then make sure these start with `s3://`, otherwise file names are assumed to be locally stored. If the `output_url` argument is to a local location, then make sure it begins with `file://`, otherwise it is assumed to be from a database.
 
 ### Method 4.
 
-If you would like to run the parser for the latest scraped files and to save the output locally, then run
-`python parse_latest.py organisation`. Where `organisation` is one of gov_uk, msf, nice, parliament, unicef, or who_iris.
+If you would like to run the parser for the latest scraped files and to save the output locally, then first save an environmental variable with your organisation name in and then run the parser, e.g. for msf:
+```
+export ORGANISATION=msf
+python parse_latest.py msf
```
Comment (nsorros) [inline README.md]: This should be in a separate PR but let's leave it in for now.

File: parse_latest.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nsorros) [reply parse_latest.py]: The better way to do this is 'scraper-results/who_iris/20181204.json'.split('.')[0].split('/')[1]. An even better way would be to use a regex.

File: refparse.py [hunk 18]
```diff
@@ -176,6 +177,106 @@ def run_predict(scraper_file, references_file,
...
+            default (number of cores).
+        logger: logging configuration name
+    """
+    print(type(num_workers))
+    if num_workers == 1:
+
+        pool = None
+        pool_map = map
+    else:
+        print("here")
```
Comment (nsorros) [inline refparse.py]: remove

File: refparse.py [hunk 19]
```diff
@@ -176,6 +177,106 @@ def run_predict(scraper_file, references_file,
...
+        pool = Pool(num_workers)
+        pool_map = pool.map
+        # logger.info(
```
Comment (nsorros) [inline refparse.py]: remove comments

File: refparse.py [hunk 20]
```diff
@@ -176,6 +177,106 @@ def run_predict(scraper_file, references_file,
     )
 
 
+def parse_references(scraper_file, references_file,
+    model_file, vectorizer_file, output_url, num_workers, logger):
```
Comment (nsorros) [inline refparse.py]: is adding the logger into the arguments a common pattern? @hblanks

File: parse_latest.py [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+
+    # The directory name will be the name of the organisation 
+    # and the date of scrape (which is the name of the file)
+    folder_name = org + os.path.splitext(os.path.basename(key_name))[0]
+    dir_name = '{}_{}'.format(args.output_url,folder_name)
+
+    if not os.path.exists(dir_name[7:]):
+        # use everything after first two slashes;
+        os.mkdir(dir_name[7:])
```
Comment (nsorros) [inline parse_latest.py]: is the above code intention to create a special place to save the output?

File: README.md (hunk 17 above)
Comment (lizgzil) [reply README.md]: 😎

File: refparse.py (hunk 18 above)
Comment (lizgzil) [reply refparse.py]: 👍

File: refparse.py (hunk 19 above)
Comment (lizgzil) [reply refparse.py]: 👍

File: parse_latest.py (hunk 21 above)
Comment (lizgzil) [reply parse_latest.py]: yes

File: refparse.py [hunk 22]
```diff
@@ -176,6 +174,100 @@ def run_predict(scraper_file, references_file,
...
+        scraper_file: path / S3 url to scraper results file
+        references_file: path/S3 url to references CSV file
+        model_file: path/S3 url to model pickle file (three formats FTW!)
+        vectorizer_file: path/S3 url to vectorizer pickle file
+        output_url: file/S3 url for output files
+        num_workers: number of workers to use, or None for multiprocessing's
+            default (number of cores).
+        logger: logging configuration name
+    """
+    print(type(num_workers))
```
Comment (nsorros) [inline refparse.py]: remove

File: parse_latest.py (hunk 21 above)
Comment (nsorros) [reply parse_latest.py]: do not use `dir_name[7:]` to split, there is probably a better way to access whatever you need to access. happy to chat about it.

File: refparse.py (hunk 22 above)
Comment (lizgzil) [reply refparse.py]: 👍

**Comment by lizgzil** [comment]
@nsorros latest commit with the changes we did together
